### PR TITLE
Minor fix for aux joomla_registration_privesc

### PR DESCRIPTION
**Pull request by aushack**
Module checks for version number. If disclosed, checks for vulnerable release ranges.
Missing a Safe retval for known immune numbers.

Fixes #7498.

### CODE REVIEW

**Comment by h00die** [comment]
Instead of the broken out statement, what about just adding `Exploit::CheckCode::Safe` after the old return statement?
I believe `Exploit` is also not required.  This way if its valid it returns, if not the CheckCode::Safe will be the last thing, and thus returned

**Comment by aushack** [comment]
I'm used to `Exploit` ?!

Fair point, I don't have the means to check exhaustively. The original code would appear to check for a responsive CMS instance and return `Detected` if present. The vulnerability is specific to known versions, so the current logic will report `Detected` where the version is known to be immune. Hence added the `Safe` return, whilst leaving the `Detected` where the module may fingerprint the CMS without knowing specific versions (I'm half guessing).

**Comment by bcoles** [comment]
`Exploit` is redundant in `Exploit` context; however, I believe it is required in `Auxiliary` context.

File: modules/auxiliary/admin/http/joomla_registration_privesc.rb [hunk 1]
```diff
@@ -58,7 +58,11 @@ def check
...
       print_status("Detected Joomla version #{joomla_version}")
-      return Exploit::CheckCode::Appears if version.between?(Gem::Version.new('3.4.4'), Gem::Version.new('3.6.3'))
+      if version.between?(Gem::Version.new('3.4.4'), Gem::Version.new('3.6.3'))
+        return Exploit::CheckCode::Appears
+      else
+        return Exploit::CheckCode::Safe
+      end
     end
 
     return Exploit::CheckCode::Detected if online
```
Comment (bcoles) [inline modules/auxiliary/admin/http/joomla_registration_privesc.rb]: I know it's not part of the changes in this PR, but this happy little conditional `if online`  should always be `true`, by virtue of bailing out earlier with `unless online`, rendering it pointless.

```suggestion
    return Exploit::CheckCode::Detected
```

This also makes allocating `online` as a variable a few lines above totally pointless too.

**Comment by bcoles** [comment]
It would be nice to `vprint` that the version wasn't identified.

Combining @h00die 's comment with my comments above, and adding some consistency for capitalization of `Joomla`, results in the following `check` method:

```ruby
  def check
    res = send_request_cgi('uri' => target_uri.path)

    unless res
      print_error("Connection timed out")
      return Exploit::CheckCode::Unknown
    end

    unless joomla_and_online?
      print_error("Unable to detect Joomla version")
      return Exploit::CheckCode::Safe
    end

    version = Gem::Version.new(joomla_version)

    unless version
      print_error('Unable to detect Joomla version')
      return Exploit::CheckCode::Detected
    end

    print_status("Detected Joomla version #{joomla_version}")

    if version.between?(Gem::Version.new('3.4.4'), Gem::Version.new('3.6.3'))
      return Exploit::CheckCode::Appears
    end

    Exploit::CheckCode::Safe
  end
```

**Comment by wvu-r7** [comment]
I should be able to repro this as in #7498.

**Comment by wvu-r7** [comment]
It looks like the `check` method needed more work. I'm refactoring it now.

**Comment by wvu-r7** [comment]
```
msf5 auxiliary(admin/http/joomla_registration_privesc) > check

[*] Detected Joomla version 3.4.5
[*] 127.0.0.1:8080 - The target appears to be vulnerable.
msf5 auxiliary(admin/http/joomla_registration_privesc) >
```

**Comment by wvu-r7** [comment]
# Release Notes

This fixes and refactors the `check` method in `auxiliary/admin/http/joomla_registration_privesc` to perform as intended.

**Pull request merged**